src/main/kotlin/org/rust/lang/core/types/infer/outlives/TransitiveRelation.kt

```diff
@@ -0,0 +1,196 @@
+        return result
+    }
+
+    private fun computeClosure(): BitMatrix {
+        val matrix = BitMatrix(elements.size, elements.size)
+        var changed = true
+        while (changed) {
+            changed = false
+            for ((source, target) in edges) {
+                // add an edge from S -> T
+                changed = changed or matrix.set(source, target)
+
+                // add all outgoing edges from T into S
+                changed = changed or matrix.set(target, source)
+            }
+        }
+        return matrix
+    }
+}
+
+/**
+ * Pare down is used as a step in the LUB computation. It edits the candidates array in place by removing any element
+ * j for which there exists an earlier element `i < j` such that `i -> j`.
+ */
+private fun pareDown(candidates: MutableList<Int>, closure: BitMatrix) {
+    var i = 0
+    while (i < candidates.size) {
+        val candidate1 = candidates[i++]
+        var j = i
+        var dead = 0
+        while (j < candidates.size) {
+            val candidate2 = candidates[j]
+            if (closure.get(candidate1, candidate2)) {
+                // If `i` can reach `j`, then we can remove `j`.
+                // So just mark it as dead and move on; subsequent indices will be shifted into its place.
+                dead++
+            } else {
+                candidates[j - dead] = candidate2
+            }
+            j++
+        }
+        candidates.dropLast(dead)
+    }
+}
+
+/**
+ * A "bit matrix" is basically a matrix of booleans represented as one gigantic bitset.
+ * In other words, it is as if you have [rows] bitsets, each of length [columns].
+ */
+class BitMatrix(private val rows: Int, private val columns: Int) {
+    private val bits: BitSet = BitSet(rows * columns)
+
+    /**
+     * Sets the cell at ([row], [column]) to true.
+     * Put another way, add [column] to the bitset for [row].
+     * @return true if this changed the matrix, and false otherwise.
+     */
+    fun set(row: Int, column: Int): Boolean {
+        require(row < rows && column < columns)
+        val oldValue = get(row, column)
+        bits.set(getIndex(row, column))
+        return oldValue
+    }
+
+    /**
+     * Do the bits from [row] contain [column]? Put another way, is the matrix cell at ([row], [column]) true?
+     * Put yet another way, if the matrix represents (transitive) reachability, can [row] reach [column]?
+     */
+    fun get(row: Int, column: Int): Boolean {
+        require(row < rows && column < columns)
+        return bits.get(getIndex(row, column))
+    }
+
+    /** @return those indices that are true in [row1] and [row2]. */
+    fun intersection(row1: Int, row2: Int): MutableList<Int> {
+        require(row1 < rows && row2 < rows)
+        val bits1 = bits.get(getIndex(row1), getIndex(row1 + 1))
+        val bits2 = bits.get(getIndex(row2), getIndex(row2 + 1))
+        bits1.and(bits2)
+        return bits1.stream().toList().toMutableList()
```

Comment:
Maybe we implement own conversion to list for `BitSet`? Because current code copies bit set two times

Reply:
We can use `toSequence` instead of `toList` to avoid that, I guess